python/ray/node.py

```diff
@@ -0,0 +1,507 @@
+                           check_alive=True,
+                           wait=False):
+        process_infos = self.all_processes[process_type]
+        if process_type != PROCESS_TYPE_REDIS_SERVER:
+            assert len(process_infos) == 1
+        for process_info in process_infos:
+            p = process_info.process
+            # Handle the case where the process has already exited.
+            if p.poll() is not None:
+                if check_alive:
+                    raise Exception("Attempting to kill a process of type "
+                                    "'{}', but this process is already dead."
+                                    .format(process_type))
+                else:
+                    continue
+
+            if process_info.use_valgrind:
+                p.terminate()
+                p.wait()
+                if p.returncode != 0:
+                    raise Exception("Valgrind detected some errors.")
+                continue
+
+            if process_info.use_profiler:
+                # Give process signal to write profiler data.
+                os.kill(p.pid, signal.SIGINT)
+                # Wait for profiling data to be written.
+                time.sleep(0.1)
+
+            if allow_graceful:
+                # Allow the process one second to exit gracefully.
+                p.terminate()
+                timer = threading.Timer(1, lambda p: p.kill(), [p])
+                try:
+                    timer.start()
+                    p.wait()
+                finally:
+                    timer.cancel()
+
+                if p.poll() is not None:
+                    continue
+
+            # If the process did not exit within one second, force kill it.
+            p.kill()
+            # The reason we usually don't call p.wait() here is that there's
+            # some chance we'd end up waiting a really long time.
+            if wait:
+                p.wait()
+
+        del self.all_processes[process_type]
+
+    def kill_redis(self, check_alive=True):
+        self._kill_process_type(
+            PROCESS_TYPE_REDIS_SERVER, check_alive=check_alive)
+
+    def kill_plasma_store(self, check_alive=True):
+        self._kill_process_type(
+            PROCESS_TYPE_PLASMA_STORE, check_alive=check_alive)
+
+    def kill_raylet(self, check_alive=True):
+        self._kill_process_type(PROCESS_TYPE_RAYLET, check_alive=check_alive)
+
+    def kill_log_monitor(self, check_alive=True):
+        self._kill_process_type(
+            PROCESS_TYPE_LOG_MONITOR, check_alive=check_alive)
+
+    def kill_monitor(self, check_alive=True):
+        self._kill_process_type(PROCESS_TYPE_MONITOR, check_alive=check_alive)
+
+    def kill_raylet_monitor(self, check_alive=True):
+        self._kill_process_type(
+            PROCESS_TYPE_RAYLET_MONITOR, check_alive=check_alive)
+
+    def kill_all_processes(self, check_alive=True):
+        # TODO(rkn): This is slower than necessary because it calls kill, wait,
```

Comment:
One thing we can do is to create a "killing_process_list" to store processes we have sent killing signals to, and we use another function to wait all these processes when `wait=True`. This also allows us to track the status of these being killed processes by checking the "killing_process_list".